File: JanperCab/3-Application/Dtos/DeliveryRunSheet/DeliveryRunSheetForListDto.cs
using _3_Application.Dtos.Driver;
using _3_Application.Dtos.Enquiry;
using _3_Application.Dtos.Truck;
using System.Collections.Generic;

namespace _3_Application.Dtos.DeliveryRunSheet
{
    public class DeliveryRunSheetForListDto : DeliveryRunSheetDto
    {
        public DriverDto Driver { get; set; }

        public TruckDto Truck { get; set; }

        public List<EnquiryForRunSheetDto> EnquiriesForRunSheet { get; set; }
    }
}

File: JanperCab/5-JanperCab/ClientApp/src/app/_models/delivery-run-sheet/DeliveryRunSheetForListDto.ts
import { DeliveryPatchDto } from './DeliveryPatchDto';
import { DeliveryRunSheetDto } from './DeliveryRunSheetDto';
import { DriverDto } from './../driver/DriverDto';
import { EnquiryForRunSheetDto } from './../enquiry/EnquiryForRunSheetDto';
import { Type } from 'class-transformer';

export class DeliveryRunSheetForListDto extends DeliveryRunSheetDto {
  @Type(() => DriverDto)
  driver: DriverDto;

  @Type(() => EnquiryForRunSheetDto)
  enquiriesForRunSheet: EnquiryForRunSheetDto[];

  getPatchDetails = (): DeliveryPatchDto[] => {
    const patches: DeliveryPatchDto[] = [];

    this.enquiriesForRunSheet.forEach((enquiry) => {
      const existPatch = patches.find((p) => p.hasSameAddress(enquiry));

      if (existPatch) {
        existPatch.enquiriesForRunSheet.push(enquiry);
      } else {
        patches.push(new DeliveryPatchDto(enquiry));
      }
    });

    return patches;
  };
}
